### PR TITLE
create AuthorizationPolicy using VirtualService

### PR DESCRIPTION
**Pull request by rahulanand16nov**
This PR:
- Creates AuthorizationPolicy using VirtualService and an annotation stating the CUSTOM extension provider.
- Deletes the policy if VirtualService is deleted or Annotation is removed.

Notes:
- Regex ([StringMatch](https://istio.io/latest/docs/reference/config/networking/virtual-service/#StringMatch)) is not supportable as AuthPolicy only supports: exact, prefix, and suffix. So, right now, only taking care of exact and prefix StringMatch types.
- Annotation changed from `kuadrant.io/auth-enabled: true` to `kuadrant.io/auth-provider: "sample-extz-auth-provider"`

## Steps to Verify
1. Add extension provider into the mesh configuration. https://istio.io/latest/docs/tasks/security/authorization/authz-custom/
2. `k apply -f examples/toystore/toystore.yaml`
3. `make install && make run`
4. Add `auth-provider` annotation to the virtualservice (`examples/toystore/virtualservice.yaml`)
5. Verify generated authorizationPolicy

### CODE REVIEW

File: controllers/apim/virtualservice_controller.go
Comment (maleck13) [inline controllers/apim/virtualservice_controller.go:78]: I don't think this needs to be an error? Perhaps we just log and return?

Comment (rahulanand16nov) [reply controllers/apim/virtualservice_controller.go]: Yea, we can just log it. updated.

File: controllers/apim/virtualservice_controller.go
Comment (maleck13) [inline controllers/apim/virtualservice_controller.go:109]: this is controlling which paths and methods go to Authorino? Do we need that level or for now should we just send any request to the host?

Comment (rahulanand16nov) [reply controllers/apim/virtualservice_controller.go:109]: yes, specific paths. Better to specific I guess?

File: controllers/apim/virtualservice_controller.go
Comment (maleck13) [inline controllers/apim/virtualservice_controller.go:155]: why do we need to add this? Is it due to a difference between VS and AuthorizationPolicy? Does prefix always end with ```/``` or is it valid to have ```/cars*```?

Comment (rahulanand16nov) [reply controllers/apim/virtualservice_controller.go:155]: Yes, due to differences between them. From the docs `Prefix match: “abc*” will match on value “abc” and “abcd”.`, it looks like it will be valid.

File: controllers/apim/virtualservice_controller.go
Comment (maleck13) [inline controllers/apim/virtualservice_controller.go:190]: this feels weird in a map function should we not just return the reconcile request and handle this in the reconcile logic?

Comment (rahulanand16nov) [reply controllers/apim/virtualservice_controller.go:190]: Creating a request for it means creating for all the events. Shouldn't we only care about virtualservices with annotation in the reconcile logic?

Comment (rahulanand16nov) [reply controllers/apim/virtualservice_controller.go:190]: I can move it if you want?

**Review by eguzki — CHANGES_REQUESTED** [review]

File: controllers/apim/virtualservice_controller.go
Comment (eguzki) [reply controllers/apim/virtualservice_controller.go:211]: Check the services controller of the current main branch. Use a label instead of annotation for builtin filtering out of events. https://github.com/Kuadrant/kuadrant-controller/blob/main/controllers/service_controller.go#L341

```go
return ctrl.NewControllerManagedBy(mgr).                                                                   
    For(&corev1.Service{}, builder.WithPredicates(kuadrantServicesPredicate())).                           
    Owns(&networkingv1beta1.API{}).                                                                        
    WithLogger(log.Log). // use base logger, the manager will add prefixes for watched sources             
    Complete(r) 
```

**Review by eguzki — CHANGES_REQUESTED** [review]

File: controllers/apim/virtualservice_controller.go
Comment (rahulanand16nov) [reply controllers/apim/virtualservice_controller.go:211]: `For` and `Owns`, confused me a bit. But it makes sense since AuthorizationPolicy is being generated. But I am not sure we need that `WithPredicated` since filtering is the same for all the events.

File: controllers/apim/virtualservice_controller.go
Comment (rahulanand16nov) [reply controllers/apim/virtualservice_controller.go:109]: This behavior is similar to AuthPolicy generated using OAS.

File: controllers/apim/virtualservice_controller.go
Comment (guicassolato) [reply controllers/apim/virtualservice_controller.go:190]: I agree with @maleck13 . It looks like this is changing the state of the cluster while filtering the resources.

**Comment by guicassolato** [comment]
About step number 1 to verify:

> Add extension provider into the mesh configuration. https://istio.io/latest/docs/tasks/security/authorization/authz-custom/

If you deploy with https://github.com/Kuadrant/kuadrantctl, an extension provider should be defined with name `kuadrant-authorization` already:

https://github.com/Kuadrant/kuadrantctl/blob/578161bea609a0ed053e2ce6f457911a257a10b6/istiomanifests/autogenerated/Base/Pilot/Pilot.yaml#L190

File: controllers/apim/virtualservice_controller.go
Comment (rahulanand16nov) [reply controllers/apim/virtualservice_controller.go]: Filter is now done in the reconcile logic along with the orphan check.

File: controllers/apim/virtualservice_controller.go
Comment (rahulanand16nov) [reply controllers/apim/virtualservice_controller.go]: Thanks, @eguzki! looks much more simpler now with `For` and `Owns` :)

**Comment by rahulanand16nov** [comment]
@guicassolato so it's in fact automatically added, nice!

Sample of generated AuthorizationPolicy:
```yaml
apiVersion: security.istio.io/v1beta1
  kind: AuthorizationPolicy
  metadata:
    creationTimestamp: "2022-01-31T12:26:56Z"
    generation: 1
    name: source-virtualservice-toystore
    namespace: kuadrant-system
    ownerReferences:
    - apiVersion: networking.istio.io/v1alpha3
      kind: VirtualService
      name: toystore
      uid: 4ce46ede-9775-4a16-a422-e3297f9018db
    resourceVersion: "5523"
    uid: a2badcc7-b1f3-4298-b1ce-d6ec56b55805
  spec:
    action: CUSTOM
    provider:
      name: sample-extz-auth-provider
    rules:
    - to:
      - operation:
          hosts:
          - '*.toystore.com'
          methods:
          - GET
          paths:
          - /getToy
      - operation:
          hosts:
          - '*.toystore.com'
          methods:
          - POST
          paths:
          - /admin/addToy
```

File: controllers/apim/virtualservice_controller.go
Comment (rahulanand16nov) [reply controllers/apim/virtualservice_controller.go]: I remembered that I added the ownerRef to the AuthPolicy so, in theory, orphan authpolicy shouldn't exist at all. I'll just remove that check :thinking:

File: controllers/apim/virtualservice_controller.go
Comment (rahulanand16nov) [reply controllers/apim/virtualservice_controller.go]: Hmm, I see :thinking:.... Updated.

File: controllers/apim/virtualservice_controller.go
Comment (rahulanand16nov) [reply controllers/apim/virtualservice_controller.go]: Done.

Comment (rahulanand16nov) [reply controllers/apim/virtualservice_controller.go]: Sorry, my bad, I added orphan check because let's say the user removes the annotation, and there is no automatic removal of AuthPolicy. I re-added the check.

**Comment by rahulanand16nov** [comment]
@maleck13 @eguzki @guicassolato Approvals before I merge?

**Review by maleck13 — COMMENTED** [review]
This looks pretty good to me @eguzki if you are happy lets go ahead and merge

**Comment by eguzki** [comment]
I do not want to block the PR any longer, so it gets my LGTM.

However, there is  one improvement  I wanted to highlight.

Filter events to reconcile only Virtualservice objects with kuadrant labels. Currently, any change on any virtualservice of the cluster will generate reconciliation loop and at least two GET operations against the k8s API server. Adding predicates (like the [service controller implemented here](https://github.com/Kuadrant/kuadrant-controller/blob/main/controllers/service_controller.go#L343))  to the watcher not only avoids unwanted reconciliation loops, also makes the code easier as you can assume the event is coming from a kuadrant enabled virtual service. And if it does not have the label, you know they just removed it and your duty is to delete the managed authorizationpolicy.

**Review by eguzki — APPROVED** [review]

**Comment by rahulanand16nov** [comment]
> I do not want to block the PR any longer, so it gets my LGTM.
> 
> However, there is one improvement I wanted to highlight.
> 
> Filter events to reconcile only Virtualservice objects with kuadrant labels. Currently, any change on any virtualservice of the cluster will generate reconciliation loop and at least two GET operations against the k8s API server. Adding predicates (like the [service controller implemented here](https://github.com/Kuadrant/kuadrant-controller/blob/main/controllers/service_controller.go#L343)) to the watcher not only avoids unwanted reconciliation loops, also makes the code easier as you can assume the event is coming from a kuadrant enabled virtual service. And if it does not have the label, you know they just removed it and your duty is to delete the managed authorizationpolicy.

@eguzki I wanted to implement that but was not sure how to delete the AuthorizationPolicy. When users delete the label, how can the controller know which AuthPolicy to delete if it doesn't reach the Reconcile logic (filtered before enqueuing)?

**Comment by eguzki** [comment]
Deletion should not be done in the mapper. I agree. But, if you look at the predicate implementation linked previously, the Update function checks for the label on the `old` object. Even if the label was deleted, the predicate tells the manager to get the event and run reconciliation event. In the reconciler,  check for the existence of the label. If it is not there, you know it used to be there but not anymore, hence proceed with the deletion of the owned object. If you happen to know de name of the authorizationpolicy, even easier, just delete the object.

**Review by guicassolato — COMMENTED** [review]
Tested locally, with the kuadrant-controller image built and pushed to the cluster (instead of `make run`), rolled out on top of the default stable image deployed by `kuadrantctl install`.

I didn't have to define the extension provider into the istio configmap  as it's set by default by kuadrantctl. All I had to do was:
1. apply `examples/toystore/toystore.yaml`
2. apply `examples/toystore/virtualservice.yaml` with added annotation `kuadrant.io/auth-provider: kuadrant-authorization`
3. create an `AuthConfig` + API key secret for host `api.toystore.com`
4. `kubectl -n kuadrant-system port-forward services/kuadrant-gateway 8000:80 &`
5. `curl -H 'Host: api.toystore.com' -H 'Authorization: APIKEY ...' http://localhost:8000/getToy # 200`

Then I tried:
- `kubectl -n kuadrant-system delete virtualservices/toystore` → deletes the `AuthorizationPolicy` and requests to the API start returning 404
- Removing the annotation from the virtual service → deletes the `AuthorizationPolicy` but requests to the API continue to be served – i.e. the kuadrant-gateway continues to check with Authorino, which still knows the host given the `AuthConfig` is still there.

**Comment by rahulanand16nov** [comment]
@eguzki Added predicates.

@guicassolato Looks expected behavior except the last `the kuadrant-gateway continues to check with Authorino, which still knows the host given the AuthConfig is still there.`, right? But I don't feel it's a problem with this PR?

**Comment by rahulanand16nov** [comment]
Thanks guys for the amazing feedback! Looks good enough to merge for now.

**Comment by rahulanand16nov** [comment]
Reference AuthConfig and secret for easy testing for the future:
```
kubectl -n kuadrant-system apply -f -<<EOF
apiVersion: authorino.kuadrant.io/v1beta1
kind: AuthConfig
metadata:
  name: simple-api-protection
spec:
  hosts:
  - api.toystore.io
  identity:
  - name: friends
    apiKey:
      labelSelectors:
        group: friends
    credentials:
      in: authorization_header
      keySelector: APIKEY
EOF
```
```
kubectl -n kuadrant-system apply -f -<<EOF
apiVersion: v1
kind: Secret
metadata:
  name: api-key
  labels:
    authorino.kuadrant.io/managed-by: authorino
    group: friends
stringData:
  api_key: SUPER_SECRET_STUFF
type: Opaque
EOF
```

**Pull request merged**
